### PR TITLE
Fix 'ax' and 'figwidth' options to mpl drawer

### LINKED ISSUE #6367: Plotting a circuit with matplotlib interferes with global figure
### Information

- **Qiskit Terra version**: 0.18.0.dev0+98b4a1f
- **Python version**: 3.8
- **Operating system**: Windows 10

### What is the current behavior?

Plotting a circuit diagram with matplotlib resizes the figure window

### Steps to reproduce the problem

Here is a minimal example. We create a matplotlib window with a specified size and layout and we require qiskit to draw into the specified axis `ax2`. 
```
import matplotlib.pyplot as plt
from qiskit import QuantumCircuit

Fig=plt.figure(1, figsize=(4,6))
plt.clf()
ax1=Fig.add_subplot(1,2,1)
ax2=Fig.add_subplot(1,2,2)
ax1.plot([1,2,3], [4,7,4])
print(Fig.get_size_inches())

circ = QuantumCircuit(2, name='test')
for ii in range(10):
    circ.h(1)
circ.cz(0,1)
circ.draw(ax=ax2, output='mpl')
print(Fig.get_size_inches())
```
After plotting the figure window has been resized.

### What is the expected behavior?

Plotting on a specified axis should not interface with the other axis on a figure and the figure itself.

### Suggested solutions

The problem is at this line: https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/visualization/matplotlib.py#L912
There the global figure properties are updated.

### PR DESCRIPTION
**Pull request by enavarro51**
### Summary

Fixes #6367

### Details and comments

The 'mpl' drawer has 2 options, `ax` and `figwidth`, both of which scale the drawing to a size in inches. In the `ax` case, a matplotlib subplot is first created and then the `ax` that is returned is passed to the drawer. The circuit drawing can then be placed in a subplot next to a graph, for example. An image of this can be seen in #6367. The `figwidth` option to the `style` dict sets a width for the drawing in inches.

Neither of these options work in the current 'mpl' drawer. The problem is that though parts of the drawing will scale automatically, fonts and line widths must be multiplied by a scaling factor. But the size of the scaling factor cannot be known until the drawing is rendered. The only option given the current 'mpl' structure would be to draw the circuit twice, once to get the scaling factor and a second time to scale the drawing to the number of inches.

This PR restructures the 'mpl' drawer. Now all the information to draw a circuit is first saved to a '_data' dict keyed per node. This allows for the computation of the finished size of the drawing without calling any matplotlib drawing functions. It also simplifies the code, so now for example a `self._gate` call is simply `self._gate(node)`.

There are now 4 ways to scale a drawing, in order of priority,
- Enter `scale` kwarg to the `draw` command as a float ratio, where 1.0 is the default scaling.
- Create a matplotlib subplot and pass the `ax` that's returned as a kwarg to the `draw` command.
- Set the `figwidth` option in the `style` dict to set a width in inches.
- Do nothing and get the default size.

There is about a 5% decrease in load time with these changes. There is an increase in memory usage, but it should be no more than 200k for 1000 gates. It's been tested up to 10,000 gates without error.

As part of this update, f-strings have replaced most of the formats, and gate `name` is no longer used to determine gate `type`, so `op.name == 'x'` has been replaced by `isinstance(op, XGate)`.

The order of methods in the class have been changed to improve readability and maintenance. The `draw` method is the entry and exit point for the class, so it has been moved to the top. Methods called from `draw` now follow in the order they are called.

### CODE REVIEW

**Review by javabster — COMMENTED** [review]
Thanks for all this Ed, I think it's generally looking good! Could you please add a release note?

File: qiskit/visualization/matplotlib.py [hunk 1]
```diff
@@ -302,34 +259,354 @@ def __init__(
...
+        self._ax.set_xlim(xl, xr)
+        self._ax.set_ylim(yb, yt)
+
+        # update figure size and, for backward compatibility,
+        # need to scale by a default value equal to (self._fs * 3.01 / 72 / 0.65)
+        base_fig_w = (xr - xl) * 0.8361111
+        base_fig_h = (yt - yb) * 0.8361111
+        scale = self._scale
+
+        # if user passes in an ax, this size takes priority over any other settings
```
[the remark this answers is not in the document]
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py:305]: Thanks, Abby. I'll get the release note done.

[the remark this answers is not in the document]
Comment (enavarro51) [reply qiskit/visualization/matplotlib.py:305]: Done in https://github.com/Qiskit/qiskit-terra/pull/6675/commits/ce548c3a4cff7b0658a0a0127237c08baadd4023

**Review by javabster — APPROVED** [review]
This generally LGTM, thanks for fixing this longstanding issue! 🚀

File: test/ipynb/mpl/circuit/test_circuit_matplotlib_drawer.py [hunk 2]
```diff
@@ -621,6 +621,33 @@ def test_overwide_gates(self):
...
+        fig.patch.set_facecolor("white")
+        fig.add_subplot(1, 2, 1)
+        ax2 = fig.add_subplot(1, 2, 2)
+
+        circuit = QuantumCircuit(4)
+        circuit.h(0)
+        circuit.cx(0, 1)
+        circuit.h(1)
+        circuit.cx(1, 2)
+        plt.close(fig)
```
Comment (1ucian0) [inline test/ipynb/mpl/circuit/test_circuit_matplotlib_drawer.py:638]: Maybe a slightly more realistic scenario?
```suggestion
        fig = plt.figure(1, figsize=(6, 4))
        ax1 = fig.add_subplot(1, 2, 1)

        circuit = QuantumCircuit(2)
        circuit.h(0)
        circuit.cx(0, 1)
        circuit.draw('mpl', ax=ax1)

        ax2 = fig.add_subplot(1, 2, 2)

        circuit = QuantumCircuit(4)
        circuit.h(0)
        circuit.cx(0, 1)
        circuit.h(1)
        circuit.cx(1, 2)
        plt.close(fig)
```

**Review by 1ucian0 — APPROVED** [review]
LGTM. Just a short comment about the subfig example. Maybe for a different PR?

**Pull request merged**